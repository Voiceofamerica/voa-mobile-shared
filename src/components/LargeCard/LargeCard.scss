
@import '../../commonStyles/index.scss';

.card {
  flex: 1;
  position: relative;
  display: flex;
  flex-direction: column;
  background-size: cover;
  background-position: center;
  overflow: hidden;
  -webkit-overflow-scrolling: touch;
  user-select: none;
  border-bottom: 4px solid;
  margin-bottom: $card-spacing;

  .overText {
    @include touchable();

    flex: 1;
    display: flex;
    flex-direction: column;
    position: relative;
    justify-content: flex-end;
    max-height: (100vw * 9 / 16);

    .imageContainer {
      position: absolute;
      top: 0;
      bottom: 0;
      left: 0;
      right: 0;
    }

    .imageStyle {
      display: block;
      width: 100%;
      height: 100%;
      object-fit: cover;
    }

    .gradient {
      display: flex;
      flex-direction: column;
      justify-content: flex-end;
      padding: 1.5vw 3.5vw 2vh;
      height: 90px;
      z-index: 1;
    }

    .titleText {
      font-weight: bold;
      font-size: 20px;
      line-height: 30px;
    }
  }

  $underTextHeight: 40px;
  $iconButtonSize: $underTextHeight - 12px;

  .underText {
    display: flex;
    flex-direction: row;
    height: $underTextHeight;
    justify-content: center;
    align-items: center;
    padding: 0 3.5vw;

    .minorTextArea {
      display: flex;
      align-items: center;
      flex: 1;
    }

    .iconButton {
      @include touchable();

      display: flex;
      justify-content: center;
      align-items: center;
      width: $underTextHeight;
      height: $underTextHeight;
      font-size: $iconButtonSize;
    }
  }

  .spinner {
    width: 1em;
    height: 1em;
  }
}
