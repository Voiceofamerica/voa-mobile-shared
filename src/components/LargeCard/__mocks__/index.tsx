
import * as React from 'react'

export default (props) => (
  React.createElement('large-card', { props }, props.children)
)

export const getHeight = () => {
  return 10
}

export const CARD_PADDING = 51
