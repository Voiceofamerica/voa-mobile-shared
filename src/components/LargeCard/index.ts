
export { default } from './LargeCard'
export * from './LargeCard'
