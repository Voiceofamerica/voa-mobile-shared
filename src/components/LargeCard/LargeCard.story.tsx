
import * as React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'

import { sleep } from '../../helpers/timingHelper'
import ThemeProvider, { BaseTheme, LargeCardTheme } from '../ThemeProvider'
import Spinner from '../Spinner'

import LargeCard from './LargeCard'

const testArticle = {
  imageUrl: 'https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w650_r1.png',
  minorText: '15:08',
  title: 'President Trump: From North Korea to Iran, we confront dangerous governments',
}

const testArticle2 = {
  imageUrl: 'https://gdb.voanews.com/01D5530E-3642-4CBF-957E-AE990ABE9167_cx0_cy14_cw0_w650_r1.jpg',
  pubDate: '21:55',
  title: 'At least seven people were killed in a police vehicle in Balochistan, Pakistan',
}

const rowStyle: React.CSSProperties = {
  display: 'flex',
  flexDirection: 'row',
  //   (9/16 image ratio)
  // + 30px of underText
  height: 'calc(56.25vw + 30px)',
}

const icons = [
  {
    icon: 'share',
    onPress: action('share iconButton.onPress'),
  },
  {
    icon: 'favorite',
    onPress: async () => {
      action('favorite iconButton pressed')()
      await sleep(1500)
      action('favirite iconButton done')()
    },
  },
  {
    icon: 'download',
    onPress: async () => {
      action('download iconButton pressed')()
      await sleep(1500)
      action('download iconButton failed')()
      throw new Error()
    },
  },
]

const testBaseTheme: Partial<BaseTheme> = {
  primaryColor: '#35A527',
  white: '#DDDDFF',
  black: '#000044',
  lightGrey: '#888888',
  darkGrey: '#111111',
  accentGrey: '#DDDDDD',
}

const testLocalTheme: Partial<LargeCardTheme> = {
  cardGradientColor: '#330000',
  cardTitleColor: '#FFDDDD',
  cardBorderColor: '#3527A5',
  largeCardUnderBackground: '#333333',
  largeCardUnderColor: '#DDDDDD',
  largeCardIconColor: '#FFFFFF',
}

storiesOf(LargeCard.name, module)
  .add('display', () => (
    <div style={rowStyle}>
      <LargeCard onPress={action('LargeCard.onPress')} { ...testArticle } />
    </div>
  ))
  .add('display with icon', () => (
    <div style={rowStyle}>
      <LargeCard
        onPress={action('LargeCard.onPress')}
        { ...testArticle }
        titleIcon='audio'
        title={(testArticle.title)}
      />
    </div>
  ))
  .add('display with icon buttons', () => (
    <div style={rowStyle}>
      <LargeCard
        onPress={action('LargeCard.onPress')}
        { ...testArticle }
        titleIcon='video'
        title={(testArticle.title)}
        iconButtons={icons}
      />
    </div>
  ))
  .add('base themed display', () => (
    <ThemeProvider value={testBaseTheme}>
      <div style={rowStyle}>
        <LargeCard
          onPress={action('LargeCard.onPress')}
          { ...testArticle }
          titleIcon='video'
          title={(testArticle.title)}
          iconButtons={icons}
        />
      </div>
    </ThemeProvider>
  ))
  .add('local themed display', () => (
    <ThemeProvider value={testLocalTheme}>
      <div style={rowStyle}>
        <LargeCard
          onPress={action('LargeCard.onPress')}
          { ...testArticle }
          titleIcon='video'
          title={(testArticle.title)}
          iconButtons={icons}
        />
      </div>
    </ThemeProvider>
  ))
  .add('rtl display with icon buttons', () => (
    <div style={rowStyle} dir='rtl'>
      <LargeCard
        onPress={action('LargeCard.onPress')}
        { ...testArticle }
        titleIcon='video'
        title={(testArticle.title)}
        iconButtons={icons}
      />
    </div>
  ))
