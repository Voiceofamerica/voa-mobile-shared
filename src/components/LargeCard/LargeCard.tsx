
import * as React from 'react'

import { sleep } from '../../helpers/timingHelper'
import { toRGBAstring } from '../../helpers/colorHelper'

import { themed, ThemeProps, DEFAULT_THEME } from '../ThemeProvider'
import ResilientImage from '../ResilientImage'
import Spinner from '../Spinner'
import SvgIcon from '../SvgIcon'

import { card, overText, imageContainer, gradient, titleText, underText, minorTextArea, iconButton, spinner } from './LargeCard.scss'

export interface IconDefinition {
  icon: string | JSX.Element
  onPress: () => Promise<void> | void
  getClassName?: () => string
  getStyle?: () => React.CSSProperties
}

export interface Props extends ThemeProps {
  title: string
  minorText?: string
  onPress?: () => void
  imageUrl?: string | null
  titleIcon?: JSX.Element | string
  className?: string
  style?: React.CSSProperties
  iconButtons?: IconDefinition[]
}

export type LoadingState = 'done' | 'loading' | 'error'
const DONE: LoadingState = 'done'
const LOADING: LoadingState = 'loading'
const ERROR: LoadingState = 'error'

export interface State {
  iconButtonsLoading: LoadingState[]
}

const IMAGE_RATIO = 9 / 16

export const getHeight = (renderWidth = window.innerWidth, imageRatio = IMAGE_RATIO) => {
  return renderWidth * imageRatio
}

// 40px under height + 7px margin-bottom
export const CARD_PADDING = 47

@themed
class LargeCard extends React.Component<Props, State> {
  state: State = {
    iconButtonsLoading: [],
  }

  private mounted: boolean

  componentDidMount () {
    this.mounted = true
  }

  componentWillUnmount () {
    this.mounted = false
  }

  render () {
    const { className, style } = this.props

    return (
      <div
        className={`${card} ${className}`}
        style={style}
      >
        {this.renderOver()}
        {this.renderUnder()}
      </div>
    )
  }

  private renderOver = () => {
    const { title, onPress, imageUrl, titleIcon, theme = DEFAULT_THEME } = this.props

    const hydratedIcon = titleIcon && this.mapIcon(titleIcon)

    const {
      cardGradientColor,
      cardTitleColor,
    } = theme

    const gradientEnd = toRGBAstring(cardGradientColor, 0.9)
    const gradientBackground = `linear-gradient(transparent, ${gradientEnd})`

    return (
      <div className={overText} onClick={onPress}>
        <ResilientImage className={imageContainer} src={imageUrl} alwaysShow showSpinner />
        <div className={gradient} style={{ background: gradientBackground}}>
          <div className={titleText} style={{ color: cardTitleColor }}>
            <span>{hydratedIcon} {title}</span>
          </div>
        </div>
      </div>
    )
  }

  private renderUnder = () => {
    const { minorText, iconButtons = [], theme = DEFAULT_THEME } = this.props
    const {
      largeCardUnderBackground: background,
      largeCardUnderColor: color,
      largeCardIconColor,
    } = theme

    return (
      <div className={underText} style={{ background, color }}>
        <div className={minorTextArea}>{minorText}</div>
        {
          iconButtons.map(({ icon, onPress, getClassName = () => '', getStyle = () => ({}) }, idx) => {
            return (
              <div key={idx} className={`${iconButton} ${getClassName()}`} onClick={this.clickHandler(idx, onPress)} style={{ color: largeCardIconColor, ...getStyle() }}>
                {this.mapIconButton(idx, icon)}
              </div>
            )
          })
        }
      </div>
    )
  }

  private mapIcon = (icon: string | JSX.Element): JSX.Element => {
    return typeof icon === 'string'
    ? <SvgIcon src={icon} />
    : icon
  }

  private mapIconButton = (idx: number, icon: string | JSX.Element): JSX.Element => {
    const { theme = DEFAULT_THEME } = this.props
    if (this.getIconButtonState(idx) === LOADING) {
      return <Spinner className={spinner} />
    } else if (this.getIconButtonState(idx) === ERROR) {
      return <SvgIcon src='close' style={{ color: theme.largeCardErrorIconColor }} />
    }

    return this.mapIcon(icon)
  }

  private clickHandler = (idx: number, originalHandler: () => Promise<void> | void) => async (): Promise<void> => {
    if (this.getIconButtonState(idx) !== DONE) {
      return
    }

    try {
      const result = originalHandler()
      await this.setIconButtonLoading(idx, LOADING)
      await result
    } catch (err) {
      await this.setIconButtonLoading(idx, ERROR)
      await sleep(1000)
    }
    await this.setIconButtonLoading(idx, DONE)
  }

  private setIconButtonLoading = (idx: number, value: LoadingState) => {
    if (!this.mounted) {
      return Promise.resolve()
    }
    return new Promise(resolve => {
      this.setState(prev => {
        const newList = prev.iconButtonsLoading.slice()
        newList[idx] = value
        return {
          ...prev,
          iconButtonsLoading: newList,
        }
      }, resolve)
    })
  }

  private getIconButtonState = (idx: number): LoadingState => {
    return this.state.iconButtonsLoading[idx] || DONE
  }
}

export default LargeCard
