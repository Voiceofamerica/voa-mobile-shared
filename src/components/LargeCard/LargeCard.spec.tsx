
import * as React from 'react'
import { create } from 'react-test-renderer'
import { mount } from 'enzyme'

import LargeCard from './LargeCard'
import { overText } from './LargeCard.scss'

jest.mock('../ThemeProvider')
jest.mock('../ResilientImage')
jest.mock('../Spinner')
jest.mock('../SvgIcon')

describe(`<${LargeCard.name} />`, () => {
  describe('shapshots', () => {
    it('should render with the given title', () => {
      const element = create((
        <LargeCard title='myTitle' />
      ))

      expect(element).toMatchSnapshot()
    })

    it('should render with the given image', () => {
      const element = create((
        <LargeCard title='myTitle' imageUrl='myImage' />
      ))

      expect(element).toMatchSnapshot()
    })

    it('should render with the given minor text', () => {
      const element = create((
        <LargeCard title='myTitle' minorText='This is some minor text' />
      ))

      expect(element).toMatchSnapshot()
    })

    it('should render with the given icon shown', () => {
      const element = create((
        <LargeCard title='myTitle' titleIcon={<i className='mdi mdi-monitor' />} />
      ))

      expect(element).toMatchSnapshot()
    })

    it('should render with the given className', () => {
      const element = create((
        <LargeCard title='myTitle' className='myClass' />
      ))

      expect(element).toMatchSnapshot()
    })

    it('should render with the given style', () => {
      const element = create((
        <LargeCard title='myTitle' style={{ top: 0, left: 0, right: 0, bottom: 0 }} />
      ))

      expect(element).toMatchSnapshot()
    })
  })

  describe('onPress', () => {
    it('should be called whenever the element is clicked', () => {
      const onPress = jest.fn()

      const item = mount((
        <LargeCard title='myTitle' onPress={onPress} />
      ))

      item.find(`.${overText}`).simulate('click')

      expect(onPress).toHaveBeenCalledTimes(1)
    })

    it('should not throw an error if not provided and the element is clicked', () => {
      const item = mount((
        <LargeCard title='myTitle' />
      ))

      item.find(`.${overText}`).simulate('click')
    })
  })
})
