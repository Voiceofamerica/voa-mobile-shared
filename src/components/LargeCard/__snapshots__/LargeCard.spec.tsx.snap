// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LargeCard /> shapshots should render with the given className 1`] = `
<div
  className="card myClass"
  style={undefined}
>
  <div
    className="overText"
    onClick={undefined}
  >
    <resilient-image
      props={
        Object {
          "alwaysShow": true,
          "className": "imageContainer",
          "showSpinner": true,
          "src": undefined,
        }
      }
    />
    <div
      className="gradient"
      style={
        Object {
          "background": "linear-gradient(transparent, rgba(0, 0, 0, 0.9))",
        }
      }
    >
      <div
        className="titleText"
        style={
          Object {
            "color": "#000000",
          }
        }
      >
        <span>
           
          myTitle
        </span>
      </div>
    </div>
  </div>
  <div
    className="underText"
    style={
      Object {
        "background": "#000000",
        "color": "#000000",
      }
    }
  >
    <div
      className="minorTextArea"
    />
  </div>
</div>
`;

exports[`<LargeCard /> shapshots should render with the given icon shown 1`] = `
<div
  className="card undefined"
  style={undefined}
>
  <div
    className="overText"
    onClick={undefined}
  >
    <resilient-image
      props={
        Object {
          "alwaysShow": true,
          "className": "imageContainer",
          "showSpinner": true,
          "src": undefined,
        }
      }
    />
    <div
      className="gradient"
      style={
        Object {
          "background": "linear-gradient(transparent, rgba(0, 0, 0, 0.9))",
        }
      }
    >
      <div
        className="titleText"
        style={
          Object {
            "color": "#000000",
          }
        }
      >
        <span>
          <i
            className="mdi mdi-monitor"
          />
           
          myTitle
        </span>
      </div>
    </div>
  </div>
  <div
    className="underText"
    style={
      Object {
        "background": "#000000",
        "color": "#000000",
      }
    }
  >
    <div
      className="minorTextArea"
    />
  </div>
</div>
`;

exports[`<LargeCard /> shapshots should render with the given image 1`] = `
<div
  className="card undefined"
  style={undefined}
>
  <div
    className="overText"
    onClick={undefined}
  >
    <resilient-image
      props={
        Object {
          "alwaysShow": true,
          "className": "imageContainer",
          "showSpinner": true,
          "src": "myImage",
        }
      }
    />
    <div
      className="gradient"
      style={
        Object {
          "background": "linear-gradient(transparent, rgba(0, 0, 0, 0.9))",
        }
      }
    >
      <div
        className="titleText"
        style={
          Object {
            "color": "#000000",
          }
        }
      >
        <span>
           
          myTitle
        </span>
      </div>
    </div>
  </div>
  <div
    className="underText"
    style={
      Object {
        "background": "#000000",
        "color": "#000000",
      }
    }
  >
    <div
      className="minorTextArea"
    />
  </div>
</div>
`;

exports[`<LargeCard /> shapshots should render with the given minor text 1`] = `
<div
  className="card undefined"
  style={undefined}
>
  <div
    className="overText"
    onClick={undefined}
  >
    <resilient-image
      props={
        Object {
          "alwaysShow": true,
          "className": "imageContainer",
          "showSpinner": true,
          "src": undefined,
        }
      }
    />
    <div
      className="gradient"
      style={
        Object {
          "background": "linear-gradient(transparent, rgba(0, 0, 0, 0.9))",
        }
      }
    >
      <div
        className="titleText"
        style={
          Object {
            "color": "#000000",
          }
        }
      >
        <span>
           
          myTitle
        </span>
      </div>
    </div>
  </div>
  <div
    className="underText"
    style={
      Object {
        "background": "#000000",
        "color": "#000000",
      }
    }
  >
    <div
      className="minorTextArea"
    >
      This is some minor text
    </div>
  </div>
</div>
`;

exports[`<LargeCard /> shapshots should render with the given style 1`] = `
<div
  className="card undefined"
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "right": 0,
      "top": 0,
    }
  }
>
  <div
    className="overText"
    onClick={undefined}
  >
    <resilient-image
      props={
        Object {
          "alwaysShow": true,
          "className": "imageContainer",
          "showSpinner": true,
          "src": undefined,
        }
      }
    />
    <div
      className="gradient"
      style={
        Object {
          "background": "linear-gradient(transparent, rgba(0, 0, 0, 0.9))",
        }
      }
    >
      <div
        className="titleText"
        style={
          Object {
            "color": "#000000",
          }
        }
      >
        <span>
           
          myTitle
        </span>
      </div>
    </div>
  </div>
  <div
    className="underText"
    style={
      Object {
        "background": "#000000",
        "color": "#000000",
      }
    }
  >
    <div
      className="minorTextArea"
    />
  </div>
</div>
`;

exports[`<LargeCard /> shapshots should render with the given title 1`] = `
<div
  className="card undefined"
  style={undefined}
>
  <div
    className="overText"
    onClick={undefined}
  >
    <resilient-image
      props={
        Object {
          "alwaysShow": true,
          "className": "imageContainer",
          "showSpinner": true,
          "src": undefined,
        }
      }
    />
    <div
      className="gradient"
      style={
        Object {
          "background": "linear-gradient(transparent, rgba(0, 0, 0, 0.9))",
        }
      }
    >
      <div
        className="titleText"
        style={
          Object {
            "color": "#000000",
          }
        }
      >
        <span>
           
          myTitle
        </span>
      </div>
    </div>
  </div>
  <div
    className="underText"
    style={
      Object {
        "background": "#000000",
        "color": "#000000",
      }
    }
  >
    <div
      className="minorTextArea"
    />
  </div>
</div>
`;
